lib/grpc/client/adapters/mint/connection_process/connection_process.ex

```diff
@@ -0,0 +1,395 @@
+        {:noreply, state}
+
+      {:ok, conn, responses} ->
+        state = State.update_conn(state, conn)
+
+        state =
+          if state.requests == %{},
+            do: state,
+            else: Enum.reduce(responses, state, &process_response/2)
+
+        check_connection_status(state)
+
+      {:error, conn, _error, _responses} ->
+        state = State.update_conn(state, conn)
+        check_connection_status(state)
+    end
+  end
+
+  @impl true
+  def handle_continue(:process_request_stream_queue, state) do
+    {{:value, request}, queue} = :queue.out(state.request_stream_queue)
+    {ref, body, _from} = request
+    window_size = get_window_size(state.conn, ref)
+    body_size = IO.iodata_length(body)
+    dequeued_state = State.update_request_stream_queue(state, queue)
+
+    cond do
+      # Do nothing, wait for server (on stream/2) to give us more window size
+      window_size == 0 -> {:noreply, state}
+      body_size > window_size -> chunk_body_and_enqueue_rest(request, dequeued_state)
+      true -> stream_body_and_reply(request, dequeued_state)
+    end
+  end
+
+  @impl true
+  def terminate(_reason, _state) do
+    :normal
+  end
+
+  defp process_response({:status, request_ref, status}, state) do
+    State.update_response_status(state, request_ref, status)
+  end
+
+  defp process_response({:headers, request_ref, headers}, state) do
+    if State.empty_headers?(state, request_ref) do
+      new_state = State.update_response_headers(state, request_ref, headers)
+
+      :ok =
+        new_state
+        |> State.stream_response_pid(request_ref)
+        |> StreamResponseProcess.consume(:headers, headers)
+
+      new_state
+    else
+      :ok =
+        state
+        |> State.stream_response_pid(request_ref)
+        |> StreamResponseProcess.consume(:trailers, headers)
+
+      state
+    end
+  end
+
+  defp process_response({:data, request_ref, new_data}, state) do
+    :ok =
+      state
+      |> State.stream_response_pid(request_ref)
+      |> StreamResponseProcess.consume(:data, new_data)
+
+    state
+  end
+
+  defp process_response({:done, request_ref}, state) do
+    :ok =
+      state
+      |> State.stream_response_pid(request_ref)
+      |> StreamResponseProcess.done()
+
+    {_ref, new_state} = State.pop_ref(state, request_ref)
+    new_state
+  end
+
+  defp chunk_body_and_enqueue_rest({request_ref, body, from}, state) do
+    {head, tail} = chunk_body(body, get_window_size(state.conn, request_ref))
+
+    case Mint.HTTP.stream_request_body(state.conn, request_ref, head) do
+      {:ok, conn} ->
+        queue = :queue.in_r({request_ref, tail, from}, state.request_stream_queue)
+
+        new_state =
+          state
+          |> State.update_conn(conn)
+          |> State.update_request_stream_queue(queue)
+
+        {:noreply, new_state}
+
+      {:error, conn, error} ->
+        if from != nil do
+          GenServer.reply(from, {:error, error})
```

Comment:
```suggestion
          # We need an explicit reply here because the process that called this GenServer
          # isn't the same one that's expecting the reply.
          GenServer.reply(from, {:error, error})
```

If it is, we can refactor this into two separate clauses of the `case` to remove some indentation and use `{:reply, ...` for one of the clauses instead.

Reply:
The way I modeled this unfortunately we can't. 

Since I had to came up with the client stream queue to avoid blow the window size error. I can't reply the caller process directly - that means reply directly on the `handle_call` ) because I need to always check the stream queue (which I use a `handle_continue` ) before I send the data to the server. 

I could refactor to a private function if you think that would be cleaner

Reply:
> The way I modeled this unfortunately we can't.
> 
> Since I had to came up with the client stream queue to avoid blow the window size error. I can't reply the caller process directly - that means reply directly on the `handle_call` ) because I need to always check the stream queue (which I use a `handle_continue` ) before I send the data to the server.
> 
> I could refactor to a private function if you think that would be cleaner

Perhaps you could return :noreply and handle what another process needs via spawn instead of a handle call. Maybe?

Reply:
not sure if I followed your suggestion.
but in here I'm returning `:noreply` and [en-queuing the request](https://github.com/beligante/grpc/blob/add-mint-adapter/lib/grpc/client/adapters/mint/connection_process/connection_process.ex#L158)

Reply:
I've looked at the code more closely now and it would be hard to follow my suggestion without having to change a lot.

Reply:
My suggestion up there still holds, though :)

Reply:
(the actual code snippet, I mean)